sqlglot/transforms.py

```diff
@@ -247,6 +247,19 @@ def remove_within_group_for_percentiles(expression: exp.Expression) -> exp.Expre
     return expression
 
 
+def add_recursive_cte_column_names(expression: exp.Expression) -> exp.Expression:
+    if isinstance(expression, exp.With) and expression.recursive:
+        for cte in expression.expressions:
+            if not cte.args["alias"].columns:
+                query = cte.this
+                if isinstance(query, exp.Union):
+                    query = query.this
+
+                cte.args["alias"].set("columns", [s.alias for s in query.selects if s.alias])
```

Comment:
what happens if there is no alias? should we make one up?

Reply:
I think we can do that yeah. Will get on it in a bit.

Reply:
FWIW the only accessible names from outside the CTE are the names in the column list, e.g.

<img width="983" alt="Screenshot 2023-05-24 at 18 37 16" src="https://github.com/tobymao/sqlglot/assets/5699893/08345e49-0a77-4fce-a99d-3f301b5e11ca">

Reply:
Right, yeah! The above was only for Trino. But you can do `select *` there too

Reply:
Done